### PR TITLE
Add operand in ternary operators rule

### PR DESCRIPTION
**Pull request by dmitrydymarchuk**
Find unnecessary ternary operators, when if and else part are equal or if and else part could be replaced by condition casted to bool.

### CODE REVIEW

**Comment by carusogabriel** [comment]
@dmitrydymarchuk This is already handled by https://github.com/slevomat/coding-standard#slevomatcodingstandardcontrolstructuresdisallowshortternaryoperator-. Also, using the short ternary operator `?:` isn't really safe, as you can hide some boolean cases inside the first member.

File: src/Rules/Operators/OperandsInTernaryOperatorRule.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	/**
+	 * @param Node $node
+	 * @param Scope $scope
+	 * @return string[]
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (!$node instanceof Ternary) {
```
Comment (carusogabriel) [inline src/Rules/Operators/OperandsInTernaryOperatorRule.php]: Unnecessary check as we are only processing `Ternary` nodes given by the `getNodeType` method return

**Comment by dmitrydymarchuk** [comment]
@carusogabriel PHPStan can be used without PHP CodeSniffer. I think phpstan-strict-rules should return error for shorthand operator in this case.

**Comment by dmitrydymarchuk** [comment]
> Ternary operator is not needed.

This part was removed

File: phpstan-baseline.neon [hunk 2]
```diff
@@ -507,6 +507,16 @@ parameters:
 			count: 1
 			path: tests/Rules/Operators/OperandsInArithmeticSubtractionRuleTest.php
 
+		-
+			message: "#^Class PHPStan\\\\Rules\\\\Operators\\\\OperandsInTernaryOperatorRuleTest extends generic class PHPStan\\\\Testing\\\\RuleTestCase but does not specify its types\\: TRule$#"
```
Comment (ondrejmirtes) [inline phpstan-baseline.neon:511]: I really don' like to add new stuff to the baseline, it should be done correctly instead.

File: src/Rules/Operators/OperandsInTernaryOperatorRule.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		if ($ifType instanceof ConstantType && $elseType instanceof ConstantType) {
+			if ($ifType->equals($elseType)) {
+				return [sprintf(
+					'If and else parts of ternary operator are equal (%s).',
+					$ifType->describe(VerbosityLevel::value())
+				)];
+			}
+			if ($ifType instanceof ConstantBooleanType && $elseType instanceof ConstantBooleanType) {
+				return $ifType->getValue()
+					? ['Ternary operator is not needed. Use just condition casted to bool.']
```
Comment (ondrejmirtes) [inline src/Rules/Operators/OperandsInTernaryOperatorRule.php:37]: This is still here.

File: src/Rules/Operators/OperandsInTernaryOperatorRule.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\ConstantType;
+use PHPStan\Type\VerbosityLevel;
+
+/**
+ * @implements \PHPStan\Rules\Rule<\PhpParser\Node\Expr\Ternary>
+ */
+class OperandsInTernaryOperatorRule implements Rule
```
Comment (ondrejmirtes) [inline src/Rules/Operators/OperandsInTernaryOperatorRule.php:16]: The rule should be added to rules.neon similarly to this: https://github.com/phpstan/phpstan-phpunit/blob/1dd916d181b0539dea5cd37e91546afb8b107e17/rules.neon#L1-L7

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by ondrejmirtes** [comment]
Hi, unfortunately I had to revert this rule (https://github.com/phpstan/phpstan-strict-rules/commit/03807e3410747ec18cd5f05a6342384adcc423bf). It finds false positives in cases like:

```php
function (\stdClass $a, \stdClass $b): void {
    rand(0, 1) ? [$a] : [$b];
};
```

Unfortunately `PHPStan\Type\Type` has no method to handle these scenarios which would be required to check whether some value is identical or not.

**Pull request merged**